### PR TITLE
FixedFramePose in optimization 2d

### PR DESCRIPTION
**Pull request by jihoonl**
Most of the code is copied from optimization_3d except the cost function replacement.

### CODE REVIEW

**Comment by drew-heenan** [comment]
Any news on this?

File: cartographer/mapping/internal/optimization/optimization_problem_2d.cc [hunk 1]
```diff
@@ -301,6 +346,53 @@ void OptimizationProblem2D::Solve(
...
+        transform::Rigid2d fixed_frame_pose_in_map;
+        if (trajectory_data.fixed_frame_origin_in_map.has_value()) {
+          fixed_frame_pose_in_map = transform::Project2D(
+              trajectory_data.fixed_frame_origin_in_map.value());
+        } else {
+          fixed_frame_pose_in_map = node_data.global_pose_2d;
+        }
+
+        C_fixed_frames.emplace(
+            std::piecewise_construct, std::forward_as_tuple(trajectory_id),
```
Comment (wohe) [inline cartographer/mapping/internal/optimization/optimization_problem_2d.cc:384]: Is the `piecewise_construct` needed here? It looks like a simple emplace would work. Or am I missing something?

File: cartographer/mapping/internal/optimization/optimization_problem_2d.cc [hunk 2]
```diff
@@ -301,6 +346,53 @@ void OptimizationProblem2D::Solve(
...
+            std::forward_as_tuple(FromPose(fixed_frame_pose_in_map)));
+        fixed_frame_pose_initialized = true;
+      }
+
+      problem.AddResidualBlock(CreateAutoDiffSpaCostFunction(constraint_pose),
+                               nullptr /* loss function */,
```
Comment (wohe) [inline cartographer/mapping/internal/optimization/optimization_problem_2d.cc:390]: Recently there were some options added to allow the use of a more suitable loss function here. This should probably be included here as well.

**Comment by wohe** [comment]
Merging this now, I'll have a look at the issues I found in the code review tomorrow.

**Pull request merged**